File: solutions/typescript/0083.ts
/**
 * 83. 删除排序链表中的重复元素
 * 给定已排序链表头，删除重复元素使每个元素只出现一次，返回已排序链表。
 * 单次扫描：curr.val === curr.next.val 时跳过后者，否则 curr 前进一步。
 *
 * 注：ListNode 类型由 LeetCode 平台注入，本文件用 untyped LNode 接口避免冲突。
 */

interface LNode { val: number; next: LNode | null }

function deleteDuplicates(head: LNode | null): LNode | null {
    let curr: LNode | null = head;
    while (curr && curr.next) {
        if (curr.val === curr.next.val) {
            curr.next = curr.next.next;
        } else {
            curr = curr.next;
        }
    }
    return head;
}

export default deleteDuplicates;

File: solutions/rust/0083.rs
/// 83. 删除排序链表中的重复元素
/// 链表已排序，提取所有值后去重相邻重复，再重建链表即可。
/// （与 82 题相同的 Rust 实现套路，避开 Box prev.next 改写难题。）
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     fn new(val: i32) -> Self {
//         ListNode { val, next: None }
//     }
// }

pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = node.next;
    }

    let mut keep: Vec<i32> = Vec::with_capacity(vals.len());
    for &v in &vals {
        if keep.last() != Some(&v) {
            keep.push(v);
        }
    }

    let mut new_head: Option<Box<ListNode>> = None;
    for &v in keep.iter().rev() {
        new_head = Some(Box::new(ListNode { val: v, next: new_head }));
    }
    new_head
}

impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        delete_duplicates(head)
    }
}